src/lavinmq/mqtt/exchange.cr

```diff
@@ -0,0 +1,133 @@
+require "../amqp/exchange"
+require "./consts"
+require "../destination"
+require "./subscription_tree"
+require "./session"
+require "./retain_store"
+
+module LavinMQ
+  module MQTT
+    class Exchange < AMQP::Exchange
+      struct BindingKey
+        def initialize(routing_key : String, arguments : AMQP::Table? = nil)
+          @binding_key = LavinMQ::BindingKey.new(routing_key, arguments)
+        end
+
+        def inner
+          @binding_key
+        end
+
+        def hash
+          @binding_key.routing_key.hash
+        end
+      end
+
+      @bindings = Hash(BindingKey, Set(MQTT::Session)).new do |h, k|
+        h[k] = Set(MQTT::Session).new
+      end
+      @tree = MQTT::SubscriptionTree(MQTT::Session).new
+
+      def type : String
+        "mqtt"
+      end
+
+      def initialize(vhost : VHost, name : String, @retain_store : MQTT::RetainStore)
+        super(vhost, name, true, false, true)
+        @body = ::IO::Memory.new
+      end
+
+      def publish(msg : Message, immediate : Bool,
+                  queues : Set(Queue) = Set(Queue).new,
+                  exchanges : Set(Exchange) = Set(Exchange).new) : Int32
+        raise LavinMQ::Exchange::AccessRefused.new(self)
+      end
+
+      def publish(packet : MQTT::Publish) : Int32
+        @publish_in_count += 1
+        headers = AMQP::Table.new({RETAIN_HEADER: true}) if packet.retain?
```

Comment:
I don't think that `NamedTuple` will be correct?

Reply:
turns out we don't need to set retain header here at all due to [MQTT-3.3.1-9]